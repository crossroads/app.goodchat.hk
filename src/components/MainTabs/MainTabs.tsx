import React from "react";
import Home from "pages/Home/Home";
import Offers from "pages/Offers/Offers";
import { Route } from "react-router";
import {
  IonLabel,
  IonRouterOutlet,
  IonTabBar,
  IonTabButton,
  IonTabs,
} from "@ionic/react";

const MainTabs = () => (
  <IonTabs>
    <IonRouterOutlet>
      <Route exact path="/offers">
        <Offers />
      </Route>
      <Route exact path="/home">
        <Home />
      </Route>
    </IonRouterOutlet>
    <IonTabBar slot="bottom">
      <IonTabButton tab="home" href="/home">
        <IonLabel>Home</IonLabel>
      </IonTabButton>
      <IonTabButton tab="offers" href="/offers">
        <IonLabel>Offers</IonLabel>
      </IonTabButton>
    </IonTabBar>
  </IonTabs>
);

export default MainTabs;
