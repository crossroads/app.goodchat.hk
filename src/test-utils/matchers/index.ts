const expectToBeOnPage = (
  container: HTMLElement,
  myPath: string,
  expectedPage: string
) => {
  const expectedPath = `/${expectedPage}`;
  expect(myPath).toEqual(expectedPath);
  expect(container.querySelector(".ion-page")).toHaveAttribute(
    "title",
    expectedPage
  );
};

export { expectToBeOnPage };
