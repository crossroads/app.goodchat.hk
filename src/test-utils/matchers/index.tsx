import userEvent from "@testing-library/user-event";
import AuthProvider from "components/AuthProvider/AuthProvider";
import useAuth from "hooks/useAuth/useAuth";
import { screen, render } from "@testing-library/react";
import { React } from "@ungap/global-this";

const expectToBeOnPage = (
  container: HTMLElement,
  myPath: string,
  expectedPage: string
) => {
  const expectedPath = `/${expectedPage}`;
  expect(myPath).toEqual(expectedPath);
  expect(container.querySelector("ion-title")).toHaveTextContent(
    new RegExp(expectedPage, "i")
  );
};

interface PageHeaderProps {
  title: string;
  privatePage: boolean;
  withBackButton: boolean;
  element: React.ReactElement;
}
const pageHeader = ({
  title,
  privatePage,
  withBackButton,
  element,
}: PageHeaderProps) => () => {
  test(`should have a ${title} title`, () => {
    const { container } = render(element);
    expect(container.querySelector("ion-header ion-title")).toHaveTextContent(
      title
    );
  });

  if (withBackButton) {
    test("should contain a back button", () => {
      const { container } = render(element);
      expect(
        container.querySelector("ion-header ion-back-button")
      ).toBeInTheDocument();
    });
  }

  if (privatePage) {
    test("should contain a log out button", () => {
      const { container } = render(element);
      expect(
        container.querySelector("ion-header ion-button")
      ).toHaveTextContent(/log out/i);
    });

    describe("log out button", () => {
      test("should log user out on click", () => {
        let isAuthenticated = true;
        const TestComponent = () => {
          isAuthenticated = useAuth().isAuthenticated;
          return element;
        };

        expect(isAuthenticated).toBe(true);

        render(
          <AuthProvider initialAuthState={true}>
            <TestComponent />
          </AuthProvider>
        );

        const logoutButton = screen.getByText(/log out/i);
        userEvent.click(logoutButton);

        expect(isAuthenticated).toBe(false);
      });
    });
  }
};

export { expectToBeOnPage, pageHeader };
