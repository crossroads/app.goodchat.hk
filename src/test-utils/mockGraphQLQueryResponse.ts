import { SetupServerApi } from "msw/lib/types/node";
import { graphql } from "msw";

const mockGraphQLQueryResponse = <T>(
  mockServer: SetupServerApi,
  query: string,
  mockResponse: T
) => {
  mockServer.use(
    graphql.query(query, (_, res, ctx) => {
      return res(ctx.data(mockResponse));
    })
  );
};

export default mockGraphQLQueryResponse;
