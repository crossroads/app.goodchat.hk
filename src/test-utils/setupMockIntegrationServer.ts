import { rest } from "msw";
import { SetupServerApi } from "msw/lib/types/node";
import { SendPinBody, VerifyBody } from "types/api_request_response";
import mockApiResponses from "test-utils/fixtures/mockApiResponses";

function setupMockIntegrationServer(mockServer: SetupServerApi) {
  const VALID_PHONE = "+85291111111";
  const VALID_PIN = "1234";
  const VALID_OTP_AUTH_KEY = "123dsfdasf";

  mockServer.use(
    rest.post<SendPinBody>(
      `${process.env.REACT_APP_API_V2_URL}/auth/send_pin`,
      (req, res, ctx) => {
        if (req.body.mobile === VALID_PHONE) {
          return res(
            ctx.status(200),
            ctx.json(mockApiResponses["auth/send_pin"].success)
          );
        } else {
          return res(
            ctx.status(422),
            ctx.json(mockApiResponses["auth/send_pin"][422].errorResponse)
          );
        }
      }
    ),
    rest.post<VerifyBody>(
      `${process.env.REACT_APP_API_V2_URL}/auth/verify`,
      (req, res, ctx) => {
        if (
          req.body.pin === VALID_PIN &&
          req.body.otp_auth_key === VALID_OTP_AUTH_KEY
        ) {
          return res(
            ctx.status(200),
            ctx.json(mockApiResponses["auth/verify"].success)
          );
        } else {
          return res(
            ctx.status(401),
            ctx.json(mockApiResponses["auth/verify"][401].errorResponse)
          );
        }
      }
    )
  );
}

export default setupMockIntegrationServer;
