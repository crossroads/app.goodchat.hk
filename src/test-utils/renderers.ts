import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { act } from "react-dom/test-utils";

/**
 * Waits for promises to resolve and state
 * updates to be applied to the DOM
 * https://reactjs.org/docs/test-utils.html#act
 */
export async function renderWithAct(
  element: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): Promise<RenderResult> {
  let renderResult: RenderResult | undefined;

  await act(async () => {
    renderResult = render(element, options);
  });

  return renderResult!;
}
