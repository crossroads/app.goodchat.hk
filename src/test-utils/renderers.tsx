import GoodChatMockedProvider from "test-utils/components/GoodChatMockedProvider/GoodChatMockedProvider";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import AuthProvider from "components/AuthProvider/AuthProvider";
import MainRouter from "components/MainRouter/MainRouter";
import { createMemoryHistory } from "history";
import { IMocks } from "@graphql-tools/mock";
import { act } from "react-dom/test-utils";
import { Router } from "react-router";
import { I18nextProvider } from "react-i18next";
import i18n from "i18n/i18n";

/**
 * Waits for promises to resolve and state
 * updates to be applied to the DOM
 * https://reactjs.org/docs/test-utils.html#act
 */
export async function renderWithAct(
  element: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): Promise<RenderResult> {
  let renderResult: RenderResult | undefined;

  await act(async () => {
    renderResult = render(element, options);
  });

  return renderResult!;
}

export type RenderPageOptions = {
  mocks? : IMocks
  authenticated?: boolean,
  history?: string[]
}

/**
 * Renders a page based on its path
 * Loads up a mock provider and creates a history
 *
 * @export
 * @param {string} path
 * @param {RenderPageOptions} [opts={}]
 * @returns
 */
export async function renderPage(path: string, opts : RenderPageOptions = {}) {
  const history = createMemoryHistory({
    initialEntries: [...(opts.history || []), path]
  });

  const renderResult = await renderWithAct(
    <I18nextProvider i18n={i18n}>
      <AuthProvider initialAuthState={opts.authenticated ?? true}>
        <GoodChatMockedProvider mockResolvers={opts.mocks || {}}>
          <Router history={history}>
            <MainRouter />
          </Router>
        </GoodChatMockedProvider>
      </AuthProvider>
    </I18nextProvider>
  );

  return {
    history,
    ...renderResult,
  };
};
