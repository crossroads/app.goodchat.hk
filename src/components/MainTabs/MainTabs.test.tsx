import React from "react";
import { render } from "@testing-library/react";
import MainTabs from "components/MainTabs/MainTabs";
import { expectToBeOnPage } from "test-utils/matchers";
import { createMemoryHistory } from "history";
import { Router } from "react-router";

const renderComponent = (initialPath: string) => {
  const history = createMemoryHistory({ initialEntries: [initialPath] });
  return {
    history,
    ...render(
      <Router history={history}>
        <MainTabs />
      </Router>
    ),
  };
};

test("renders without crashing", () => {
  const { container } = renderComponent("/home");
  expect(container).toBeInTheDocument();
});

test("renders footer correctly", () => {
  const { container } = renderComponent("/home");
  expect(container.querySelector("ion-tab-bar")).toMatchSnapshot();
});

test("visiting /home takes user to Home", () => {
  const { container, history } = renderComponent("/home");
  expectToBeOnPage(container, history.location.pathname, "home");
});

test("visiting /offers takes user to Offers", () => {
  const { container, history } = renderComponent("/offers");
  expectToBeOnPage(container, history.location.pathname, "offers");
});
