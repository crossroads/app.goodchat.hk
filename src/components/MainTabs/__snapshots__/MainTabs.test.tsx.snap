// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders footer correctly 1`] = `
<ion-tab-bar
  slot="bottom"
>
  <ion-tab-button
    href="/home"
    tab="home"
  >
    <ion-icon
      icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Home</title><path d='M261.56 101.28a8 8 0 00-11.06 0L66.4 277.15a8 8 0 00-2.47 5.79L63.9 448a32 32 0 0032 32H192a16 16 0 0016-16V328a8 8 0 018-8h80a8 8 0 018 8v136a16 16 0 0016 16h96.06a32 32 0 0032-32V282.94a8 8 0 00-2.47-5.79z'/><path d='M490.91 244.15l-74.8-71.56V64a16 16 0 00-16-16h-48a16 16 0 00-16 16v32l-57.92-55.38C272.77 35.14 264.71 32 256 32c-8.68 0-16.72 3.14-22.14 8.63l-212.7 203.5c-6.22 6-7 15.87-1.34 22.37A16 16 0 0043 267.56L250.5 69.28a8 8 0 0111.06 0l207.52 198.28a16 16 0 0022.59-.44c6.14-6.36 5.63-16.86-.76-22.97z'/></svg>"
    />
    <ion-label>
      Home
    </ion-label>
  </ion-tab-button>
  <ion-tab-button
    href="/offers"
    tab="offers"
  >
    <ion-icon
      icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>List</title><path stroke-linecap='round' stroke-linejoin='round' d='M160 144h288M160 256h288M160 368h288' class='ionicon-fill-none ionicon-stroke-width'/><circle cx='80' cy='144' r='16' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><circle cx='80' cy='256' r='16' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><circle cx='80' cy='368' r='16' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
    />
    <ion-label>
      Offers
    </ion-label>
  </ion-tab-button>
</ion-tab-bar>
`;
